src/pinocchio/setup.rs

```diff
@@ -0,0 +1,196 @@
+use crate::circuits::qap::Qap;
+use crate::math;
+use math::field_element::FieldElement as FE;
+use math::group::Group;
+type GroupType = FE;
+
+#[allow(dead_code)]
+pub struct EvaluationKey {
+    gv_ks: Vec<GroupType>,
+    gw_ks: Vec<GroupType>,
+    gy_ks: Vec<GroupType>,
+    gv_alphaks: Vec<GroupType>,
+    gw_alphaks: Vec<GroupType>,
+    gy_alphaks: Vec<GroupType>,
+    g_s_i: Vec<GroupType>,
+    g_beta: Vec<GroupType>,
+}
+
+pub struct ToxicWaste {
+    s: FE,
+    alpha_v: FE,
+    alpha_w: FE,
+    alpha_y: FE,
+    beta: FE,
+    rv: FE,
+    rw: FE,
+}
+
+impl ToxicWaste {
+    pub fn ry(self) -> FE {
+        self.rv * self.rw
+    }
+
+    pub fn sample() -> Self {
+        Self {
+            s: FE::random(),
+            alpha_v: FE::random(),
+            alpha_w: FE::random(),
+            alpha_y: FE::random(),
+            beta: FE::random(),
+            rv: FE::random(),
+            rw: FE::random(),
+        }
+    }
+}
+
+pub fn setup(qap: Qap, tw: ToxicWaste) -> EvaluationKey {
+    let (vs_mid, ws_mid, ys_mid) = (qap.v_mid(), qap.w_mid(), qap.y_mid());
+
+    let s = tw.s;
+    let alpha_v = tw.alpha_v;
+    let alpha_w = tw.alpha_w;
+    let alpha_y = tw.alpha_y;
+    let beta = tw.beta;
+    let rv = tw.rv;
+    let rw = tw.rw;
+    let ry = tw.ry();
+
+    let g = FE::generator();
```

Comment:
Maybe `GroupType::generator`?

Reply:
Yes, that seems better to understand what's going on and be able to change it later